### ISSUE TITLE
Function required to remove arbitrary tags with CSS selectors

### ISSUE THREAD

**Issue by marretjeca** [Lullabot/amp-library#96]
We use amp pages on a website that also uses a CDN. In order to maximize speed the CDN module applies dns-prefetch:

`<meta http-equiv="x-dns-prefetch-control" content="on" />
<link rel="dns-prefetch" href="//cdn.potatopro.com" />
` 

The AMP PHP Library does not process this correctly. It leaves either a <link> statement without any rel= and/or an illegal rel=  
Google considers this a fatal error and does not index the page. 

Bottom line is that  I believe the best solution is to remove this entire piece of code (think dns-prefetching is standard for amp pages anyhow).

Therefore I propose a function that strips out this specific CDN codeset. Wim Leers may be able to confirm whether the prefetch code above is typical for all use cases.

**Comment by sidkshatriya** [comment]
@marretjeca Thanks for your bug report. I'll investigate and see what might be the best way to fix this!

**Comment by sidkshatriya** [comment]
@marretjeca Instead of removing the above tags in an ad-hoc way I've done something generic. Developers can specify CSS selectors to the AMP library. If those CSS selectors match, the tags are removed from the document. This is more flexible in the long run as different users of the library will have different tags of their own that they might want removed. We are removing illegal attributes from tags but don't know of a good way to decide whether to remove the whole tag altogether (hence this new functionality).

For your specific case you can do this two ways now (after switching to the branch `user-tag-selector-blacklist` as this is not yet merged to master). 

Pick the one that is convenient to you.

**Method 1:**
Pass in options in loadHtml.

``` php
$options = [ 'remove_selectors' =>
  ['meta[http-equiv="x-dns-prefetch-control"]', 'link[rel="dns-prefetch" ']
];
$amp->loadHtml($html, $options);
...
```

Of course, this is a specific example you can specify as many or as few (or no) CSS selectors that you want to remove.

**Method 2:**
Define a function having the following name in your code. 

``` php
    function anyprefixhere_lullabot_amp_remove_selectors() {
        return ['meta[http-equiv="x-dns-prefetch-control"]', 'link[rel="dns-prefetch" '];
    }
```

Of course, this is a specific example you can specify as many or as few CSS selectors that you want to remove. You can define multiple functions as long as the function name prefix is different (think of this as a "hook").

(I suspect you'll need to use Method 2 as the Drupal module is doing the loadHtml() for you)

Try this out and let me know if this works well!

**Comment by marretjeca** [comment]
@sidkshatriya Thanks!
Where would I put this code from method 2? Also what CSS selector would I use?

**Comment by sidkshatriya** [comment]
@marretjeca 

Lets break apart the example above:

``` php
function anyprefixhere_lullabot_amp_remove_selectors() {
  return ['meta[http-equiv="x-dns-prefetch-control"]', 'link[rel="dns-prefetch" '];
}
```

You had mentioned you were having problems with:

``` html
<meta http-equiv="x-dns-prefetch-control" content="on" /> 
<link rel="dns-prefetch" href="//cdn.potatopro.com" />
```

As I understood it, you want these tags to be removed from AMP HTML.

The first tag can be selected via CSS selectors as `meta[http-equiv="x-dns-prefetch-control]` and the second tag can be selected by `link[rel="dns-prefetch"]`. Therefore the function simply returns an array of those selectors. This is in no ways unique. You could probably even do something like `head > link[rel="dns-prefetch"]` for the `<link>` tag. It all depends how specific or general you want to be. It would depend on your HTML. The syntax of this is the same syntax as one would use in jQuery to home in a specific DOM element or do CSS theming.

Regarding where to put this function, if you're using Drupal then `.module` file should be fine. The `.module` files are loaded on each Drupal request so this function will be available when called by the AMP Library. If you don't want to put it there you might have to consider autoloading implications.

Please test drive this new functionality and let me know if it works for you. You'll need to use the `user-tag-selector-blacklist` branch as this functionality is not merged to master yet

**Comment by marretjeca** [comment]
Yes, it works! thanks very much

**Comment by sidkshatriya** [comment]
@marretjeca Glad to know it worked! Closing ticket.
